lib/spack/spack/solver/asp.py

```diff
@@ -763,6 +773,11 @@ def key_fn(version):
                 version_origin_str[declared_version.origin]
             ))
 
+        for v1, v2 in zip(most_to_least_preferred, most_to_least_preferred[::-1]):
```

Comment:
I'm not sure I understand why this zip is the appropriate way to handle this -- can you explain what's going on here a bit better?

Reply:
you're right. I was trying to avoid nested loops, but I didn't think it through. This is just index reversing and not a upper.lower triangular set of matrix of possibilities.